### PR TITLE
First version of Cell-DEVS for Cadmium

### PR DESCRIPTION
**Pull request by romancardenas**
Just added models for implementing Cell-DEVS scenarios using Cadmium.

The code is fully documented, and I'm working on a document describing all the features of this new implementation.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -109,6 +109,14 @@ foreach (exampleSrc ${Examples})
     target_link_libraries(${exampleName} PUBLIC ${Boost_LIBRARIES})
 endforeach (exampleSrc)
 
+# Cell-DEVS Examples
```
Comment (sdavtaker) [inline CMakeLists.txt:112]: If you will start separating examples in groups, please move the old ones to a pdevs subdir and generalize the build.

File: example/celldevs/default/default_cell.hpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// Created by Román Cárdenas Rodríguez on 25/05/2020.
+//
+
+#ifndef CADMIUM_CELLDEVS_DEFAULT_CELL_HPP
+#define CADMIUM_CELLDEVS_DEFAULT_CELL_HPP
+
+#include <exception>
+#include <cadmium/celldevs/cell/cell.hpp>
+
+template <typename T>
```
Comment (sdavtaker) [inline example/celldevs/default/default_cell.hpp]: please call it TIME to be explicit about what to expect (if that is what you expecting)

File: example/celldevs/default/default_cell.hpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    using cadmium::celldevs::cell<T, int, int>::state;
+
+    explicit default_cell() : cadmium::celldevs::cell<T, int, int>() {}
+
+    template <typename... Args>
+    default_cell(int cell_id, int initial_state, std::vector<int> const &neighbors, std::string const &delayer_id, Args&&... args):
+            cadmium::celldevs::cell<T, int, int>(cell_id, initial_state, neighbors, delayer_id, std::forward<Args>(args)...) {}
+
+    template <typename... Args>
+    default_cell(int cell_id, int initial_state, std::unordered_map<int, int> const &vicinities, std::string const &delayer_id, Args&&... args):
```
Comment (sdavtaker) [inline example/celldevs/default/default_cell.hpp:49]: I would like to extract the initial_state from here, I also have it in the backlog to do it in the other engines. It would be nice to discuss ideas of how to do it.

File: example/celldevs/default/default_cell.hpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+#define CADMIUM_CELLDEVS_DEFAULT_CELL_HPP
+
+#include <exception>
+#include <cadmium/celldevs/cell/cell.hpp>
+
+template <typename T>
+class default_cell: public cadmium::celldevs::cell<T, int, int> {
+public:
+    using cadmium::celldevs::cell<T, int, int>::cell_id;
+    using cadmium::celldevs::cell<T, int, int>::state;
```
Comment (sdavtaker) [inline example/celldevs/default/default_cell.hpp]: It would be nice (in the examples) to comment what are those types you passing kind of:
using cadmium::celldevs::cell</* time */ T, /* state */ int, /* neightboors */ int>::cell_id;
Or give them interesting typedefs like:
using time=T; using state_t=int; using neightbors_t=int;
using cadmium::celldevs::cell<time_t, state_t, neighbots_t>

File: example/celldevs/default/default_coupled.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
+//
```
Comment (sdavtaker) [inline example/celldevs/default/default_coupled.cpp]: Please include the license description as in other files.

File: include/cadmium/celldevs/cell/cell.hpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
+/**
+* Copyright (c) 2020, Román Cárdenas Rodríguez
+* ARSLab - Carleton University
+* GreenLSI - Polytechnic University of Madrid
+* All rights reserved.
+*
+* Abstract DEVS atomic model for defining cells in Cell-DEVS scenarios
+*/
```
Comment (sdavtaker) [inline include/cadmium/celldevs/cell/cell.hpp]: Please follow standard license

File: include/cadmium/celldevs/cell/cell.hpp [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+     */
+    template <typename T, typename C, typename S, typename V=int, typename C_HASH=std::hash<C>>
+    class cell {
+    public:
+
+        using input_ports = std::tuple<typename cell_ports_def<C, S>::cell_in>;
+        using output_ports = std::tuple<typename cell_ports_def<C, S>::cell_out>;
+
+        using NV = std::unordered_map<C, V, C_HASH>;
+        using NS = std::unordered_map<C, S, C_HASH>;
```
Comment (sdavtaker) [inline include/cadmium/celldevs/cell/cell.hpp]: Whats NV? whats NS? I mostly wondering what is N, because assuming V is value and S is state.

File: include/cadmium/celldevs/cell/cell.hpp [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+
+        C cell_id;                          /// Cell ID
+        std::vector<C> neighbors;           /// Neighboring cells' IDs
+        T clock;                            /// simulation clock
```
Comment (sdavtaker) [inline include/cadmium/celldevs/cell/cell.hpp]: Is this a real-time implementation?

File: include/cadmium/celldevs/cell/cell.hpp [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+        struct state_type {
+            S current_state;                /// Cell's internal state
+            NV neighbors_vicinity;          /// Neighboring cell' vicinity type
+            NS neighbors_state;             /// neighboring cell' public state
+        };
+        state_type state;
+
+        /// A default constructor is required for compiling issues. However, it is not valid and always throws exception
+        cell(){ throw std::exception(); }
```
Comment (sdavtaker) [inline include/cadmium/celldevs/cell/cell.hpp:122]: Why? Can you explain a little? It doesn't sound right.

File: include/cadmium/celldevs/cell/cell.hpp [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+            }
+            buffer = delayer_factory<T, S>::create_delayer(delayer_id, std::forward<Args>(args)...);
+            buffer->add_to_buffer(initial_state, T());  // At t = 0, every cell communicates its state to its neighbors
+        }
+
+        /************** USER-DEFINED METHODS **************/
+        /// @return cell's next state.
+        virtual S local_computation() const { return state.current_state; }
+        /// @return delay to be applied before communicating to neighbors a new state.
+        virtual T output_delay(S const &cell_state) const { return std::numeric_limits<T>::infinity(); }
```
Comment (sdavtaker) [inline include/cadmium/celldevs/cell/cell.hpp:173]: If you expect user to define them maybe its better to mark them virtual than give default implementations, else it will be hard to debug when someone falls into the default by mistake.

File: include/cadmium/celldevs/cell/cell.hpp [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+                next_internal = buffer->next_timeout() - clock;
+            }
+            state.current_state = next;
+        }
+
+        /// Confluence transition function.
+        void confluence_transition(T e, typename cadmium::make_message_bags<input_ports>::type mbs) {
+            internal_transition();
+            external_transition(T(), move(mbs));
+        }
```
Comment (sdavtaker) [inline include/cadmium/celldevs/cell/cell.hpp:216]: Isn't this one to be expected to be fill by users?

File: include/cadmium/celldevs/cell/cell.hpp [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+        /// Time advance function.
+        T time_advance() const { return next_internal; }
+
+        /// @return the next message to be transmitted from the output delayer buffer
+        typename cadmium::make_message_bags<output_ports>::type output() const {
```
Comment (sdavtaker) [inline include/cadmium/celldevs/cell/cell.hpp:222]: Shouldm't be provided by user?
I'm really confuse about this class, you providing a lot of things that I would expect user to provide, can you comment on what is the expectation from the user point of view?

File: include/cadmium/celldevs/cell/grid_cell.hpp [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+     * @brief DEVS atomic model for defining cells in Cell-DEVS scenarios that are sorted in lattices.
+     * @tparam T the type used for representing time in a simulation.
+     * @tparam S the type used for representing a cell state.
+     * @tparam V the type used for representing a neighboring cell's vicinity. By default, it is set to integer.
+     */
+    template <typename T, typename S, typename V=int>
+    class grid_cell : public cell<T, cell_position, S, V, seq_hash<cell_position>> {
+    public:
+
+        cell_map<S, V> map;     /// Cell map
```
Comment (sdavtaker) [inline include/cadmium/celldevs/cell/grid_cell.hpp:49]: /// -> //

File: include/cadmium/celldevs/coupled/cells_coupled.hpp [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+     * @tparam S the type used for representing a cell state.
+     * @tparam V the type used for representing a neighboring cell's vicinity. By default, it is set to integer.
+     * @tparam C_HASH the hash function used for creating unordered maps with cell IDs as keys.
+     *                By default, it uses the one defined in the standard library
+     */
+    template<typename T, typename C, typename S, typename V=int, typename C_HASH=std::hash<C>>
+    class cells_coupled: public cadmium::dynamic::modeling::coupled<T> {
+    public:
+
+        using CV = std::unordered_map<C, V, C_HASH>;
```
Comment (sdavtaker) [inline include/cadmium/celldevs/coupled/cells_coupled.hpp]: whats CV? and CNV?

File: include/cadmium/celldevs/coupled/cells_coupled.hpp [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+        using CNV = std::unordered_map<C, CV, C_HASH>;
+
+        CNV vicinities_;  /// Nested unordered map: {cell ID to: {cell ID from: Vicinity between cells}}
```
Comment (sdavtaker) [inline include/cadmium/celldevs/coupled/cells_coupled.hpp]: why _ suffix? whats the meening?

File: include/cadmium/celldevs/coupled/cells_coupled.hpp [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+        void add_cell(C const &cell_id, S const &initial_state, std::vector<C> const &neighbors,
+                      std::string const &delayer_id, Args&&... args) {
+            CV vicinity = CV();
+            for (auto const &neighbor: neighbors) {
+                vicinity.insert({neighbor, V()});
+            }
+            add_cell<CELL_MODEL, Args...>(cell_id, initial_state, vicinity, delayer_id, std::forward<Args>(args)...);
+        }
+
+        /// The user must call this method right after having included all the cells of the scenario
```
Comment (sdavtaker) [inline include/cadmium/celldevs/coupled/cells_coupled.hpp:108]: /// -> //

File: include/cadmium/celldevs/delayer/transport.hpp [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    template <typename T, typename S>
+    class transport_delayer : public delayer<T, S> {
+    private:
+        /// Queue with pairs <scheduled time, state to transmit>. Messages are sorted depending on the scheduled time
+        std::priority_queue<std::pair<T, S>, std::vector<std::pair<T, S>>, std::greater<std::pair<T, S>>> delayed_outputs;
+    public:
+        transport_delayer() : delayer<T, S>() {}
+
+        /// Pushes new state to the priority queue
```
Comment (sdavtaker) [inline include/cadmium/celldevs/delayer/transport.hpp:54]: I'm starting to think you using /// with a purpose, why?

File: include/cadmium/celldevs/utils/grid_utils.hpp [hunk 18]
```diff
@@ -0,0 +1,464 @@
...
+#include <unordered_map>
+#include <algorithm>
+#include <utility>
+#include <cassert>
+#include <exception>
+#include <cmath>
+#include <boost/functional/hash.hpp>
+
+
+// Hash function for enabling sequences to be  keys of unordered maps
```
Comment (sdavtaker) [inline include/cadmium/celldevs/utils/grid_utils.hpp]: Extra space. 
"be  key"

File: include/cadmium/celldevs/utils/grid_utils.hpp [hunk 19]
```diff
@@ -0,0 +1,464 @@
+/**
```
Comment (sdavtaker) [inline include/cadmium/celldevs/utils/grid_utils.hpp:1]: This file is huge, can it be break down into common areas of work or is all unrelated stuff just put together here?

File: test/celldevs_grid_utils_test.cpp [hunk 20]
```diff
@@ -0,0 +1,80 @@
+//
```
Comment (sdavtaker) [inline test/celldevs_grid_utils_test.cpp]: I didn't look into coverage reports, but it looks like extremely small test for so much code changes.

**Review by sdavtaker — CHANGES_REQUESTED** [review]

File: include/cadmium/celldevs/cell/cell.hpp (hunk 11 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/cell.hpp:216]: Not in [Cell-DEVS](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=9004763&casa_token=4JRwIBxX-lMAAAAA:3wLEC3JuY80n-oweqfB_N1ZeWK3afLpJk0oXPUw4YhR1zEXDhZDj8vh_d_--t_7W9MYf4vpsaQ&tag=1). This formalism is on top of DEVS, and the user only has to define the local computation function (i.e., the next state of the cell depending on its previous state and its neighbors' state) and the delay function (i.e., how long the cell has to wait before communicating to its neighbors this new state). The classic DEVS functions are hidden from the user.

File: include/cadmium/celldevs/cell/cell.hpp (hunk 12 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/cell.hpp:222]: The same as above

File: include/cadmium/celldevs/cell/cell.hpp (hunk 9 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/cell.hpp:122]: I expect at least a Cell ID. Furthermore, a cell needs an initial state, its neighbors (and, if needed, their vicinity), and the type of delay to be used (either inertial or transport). However, I think that Cadmium needs the models to have a valid empty constructor to compile. That is why I just throw an exception.

File: include/cadmium/celldevs/cell/cell.hpp (hunk 8 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/cell.hpp]: It isn't. However, it keeps track of the simulated time to perform different delayed output functionalities. It is hidden to the user.

File: include/cadmium/celldevs/cell/grid_cell.hpp (hunk 13 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/grid_cell.hpp:49]: I don't know if it is standard, but at least in CLion, with ///, comments appear as docstrings. Thus, I use /// for explaining briefly the purpose of the attributes of the main classes

File: include/cadmium/celldevs/coupled/cells_coupled.hpp (hunk 16 above)
Comment (romancardenas) [reply include/cadmium/celldevs/coupled/cells_coupled.hpp:108]: Same as above. I also use /// for documenting small methods with no arguments nor return types.

File: include/cadmium/celldevs/delayer/transport.hpp (hunk 17 above)
Comment (romancardenas) [reply include/cadmium/celldevs/delayer/transport.hpp:54]: Explained above

File: include/cadmium/celldevs/cell/grid_cell.hpp (hunk 13 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/grid_cell.hpp:49]: Screen Shot 2020-06-02 at 09 20 36
Here you can see the different highlighting for // and ///

File: include/cadmium/celldevs/cell/cell.hpp (hunk 7 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/cell.hpp]: * The N comes from neighbors (neighboring cells)
* NV is an unordered map {neighbor cell id: neighbor cell vicinity}. The vicinity is the relationship between the cells (in a grid space, it may be the distance. However, the vicinity between Prof. Wainer and me is that he is my boss).
* NS is an unordered map {neighbor cell id: known neighbor cell state}. It contains the latest state of a neighboring cell.

File: include/cadmium/celldevs/coupled/cells_coupled.hpp (hunk 14 above)
Comment (romancardenas) [reply include/cadmium/celldevs/coupled/cells_coupled.hpp]: Just aliases for reducing code size.
* CV is an unordered map {neighbor cell ID: vicinity} (just as above).
* CNV is a nested unordered map {target cell ID: {neighbor cell ID: vicinity}}

Vicinities are directed (i.e., A may be neighbor of B without B being neighbor of A).

File: include/cadmium/celldevs/utils/grid_utils.hpp (hunk 19 above)
Comment (romancardenas) [reply include/cadmium/celldevs/utils/grid_utils.hpp:1]: It is a little bit "cajón de sastre", to be honest... I will work on splitting it into different files

File: include/cadmium/celldevs/coupled/cells_coupled.hpp (hunk 15 above)
Comment (romancardenas) [reply include/cadmium/celldevs/coupled/cells_coupled.hpp]: Removed

File: example/celldevs/default/default_cell.hpp (hunk 3 above)
Comment (romancardenas) [reply example/celldevs/default/default_cell.hpp:49]: I don't get it... What do you mean? Removing the initial state from the constructor?

File: test/celldevs_grid_utils_test.cpp (hunk 20 above)
Comment (romancardenas) [reply test/celldevs_grid_utils_test.cpp]: To be honest, I am not a C++ pro... I did a minimal test to ensure that the main functions work properly.
It would be nice to meet with you and discuss about these things. For example, when compiling the Cell-DEVS examples, I get a couple of warnings regarding destructors stuff, but I don't know how to deal with them. The output provided by them, as well as the step-by-step execution, appears to be OK

**Comment by romancardenas** [comment]
I have just pushed a new version addressing all your comments. Please, read through my answers and let me know any suggestion

File: example/celldevs/default/default_cell.hpp (hunk 3 above)
Comment (sdavtaker) [reply example/celldevs/default/default_cell.hpp:49]: Yes Model initialization and model definition should be independent things like it is defined in DEVS formalism. However, not something we should resolve by now. It's a common problem with most simulators.

File: include/cadmium/celldevs/cell/cell.hpp (hunk 7 above)
Comment (sdavtaker) [reply include/cadmium/celldevs/cell/cell.hpp]: Can you comment or expand the names?. They may seem obvious now, but in 6 months when someone not completely familiar with the codebase has to debug a reported issue its something that will require investigation.

File: include/cadmium/celldevs/cell/cell.hpp (hunk 8 above)
Comment (sdavtaker) [reply include/cadmium/celldevs/cell/cell.hpp]: Can you comment a little longer what is the use? Usually clock is a word very overloaded in simulation and we want to avoid someone confused about it when looking at it back in 6 months.

File: include/cadmium/celldevs/cell/cell.hpp (hunk 9 above)
Comment (sdavtaker) [reply include/cadmium/celldevs/cell/cell.hpp:122]: Ideally you should look into why a constructor is required, and if it makes no sense, remove the requirement. However, even if we do not adjust the requirements, at least put a more specific exceptiontype and a message to know where that exception came from when it is raised.

File: include/cadmium/celldevs/cell/cell.hpp (hunk 11 above)
Comment (sdavtaker) [reply include/cadmium/celldevs/cell/cell.hpp:216]: I was implementing DEVS some time ago (you can check some of its code in the basic_models/devs/) but I never finished the engine part for it. Are you going to use DEVS or PDEVS?

File: include/cadmium/celldevs/cell/grid_cell.hpp (hunk 13 above)
Comment (sdavtaker) [reply include/cadmium/celldevs/cell/grid_cell.hpp:49]: I see, nice.

File: include/cadmium/celldevs/coupled/cells_coupled.hpp (hunk 14 above)
Comment (sdavtaker) [reply include/cadmium/celldevs/coupled/cells_coupled.hpp]: We don't pay per character we use, so its fine to spend a few more letters for the sake of clarity to those coming after, there is a big chance that someone else will contribute to this code in the future, the easier it is to read, the more probable someone is willing to extend i and do not go away before finishing. :)

File: include/cadmium/celldevs/utils/grid_utils.hpp (hunk 19 above)
Comment (sdavtaker) [reply include/cadmium/celldevs/utils/grid_utils.hpp:1]: This is fine to keep for now and refactor in a follow up PR.

File: test/celldevs_grid_utils_test.cpp (hunk 20 above)
Comment (sdavtaker) [reply test/celldevs_grid_utils_test.cpp]: Sure, we can schedule a call with Laouen too. Let's do that by email.

**Comment by sdavtaker** [comment]
> I have just pushed a new version addressing all your comments. Please, read through my answers and let me know any suggestion

I replied to some of your answers.

File: include/cadmium/celldevs/cell/cell.hpp (hunk 7 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/cell.hpp]: I moved it to a shared type cell_unordered (see below)

File: include/cadmium/celldevs/cell/cell.hpp (hunk 11 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/cell.hpp:216]: I think PDEVS. Now the behavior is simple, but I am planning to add additional ports, etc.

File: include/cadmium/celldevs/cell/cell.hpp (hunk 8 above)
Comment (romancardenas) [reply include/cadmium/celldevs/cell/cell.hpp]: I changed it to simulation_clock and added a longer description

File: include/cadmium/celldevs/coupled/cells_coupled.hpp (hunk 14 above)
Comment (romancardenas) [reply include/cadmium/celldevs/coupled/cells_coupled.hpp]: Again, I removed these aliases and substituted by a template class for defining unordered maps which key is a cell ID

**Comment by romancardenas** [comment]
New commit. I have removed strange aliases and used a standard type for defining unorder maps which key is a cell ID
